### PR TITLE
[M] Migrate import_update_single_pool_spec and import_warning_spec (CANDLEPIN-364)

### PR DESCRIPTION
**Pull request by JoshAlbrecht23**
- removed import_update_single_pool_spec.rb
- removed import_warning_spec.rb

### CODE REVIEW

**Comment by Ceiu** [comment]
@bakajstep have you had a chance to look through this PR yet?

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            .singleElement()
+            .extracting(SubscriptionDTO::getProduct)
+            .isNotNull()
+            .returns(prod.getId(), ProductDTO::getId);
+
+        int expectedQuantity = 5;
+        JsonNode ent2 = adminClient.consumers()
+            .bindPool(consumer.getUuid(), subs.get(0).getUpstreamPoolId(), expectedQuantity)
+            .get(0);
+        Thread.sleep(1000);
```
Comment (bakajstep) [inline spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java:76]: I tested this on myself(in my PR with bind) so we wouldn't waste so much time sleeping. I got to 450 ms, but to be safe we can use 500 and save a second in this case.

Tested with this annotation.
```
    @RepeatedTest(30)
    @Execution(ExecutionMode.SAME_THREAD)
```

I try it in this PR too and it is working with 500.

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        export = exportGenerator.export();
+        importJob = adminClient.owners().importManifestAsync(ownerKey, List.of(), export.file());
+        importJob = adminClient.jobs().waitForJob(importJob);
+        assertEquals("FINISHED", importJob.getState());
+        subs = adminClient.owners().getOwnerSubscriptions(ownerKey);
+        assertThat(subs)
+            .hasSize(2);
+
+        adminClient.consumers().unbindByEntitlementId(consumer.getUuid(), ent1.getId());
+        Thread.sleep(1000);
```
Comment (bakajstep) [inline spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java:87]: Same here

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                        // Copy over the modified consumer_export.zip
+                        transfer(entry, temp, zos);
+                    }
+                    else {
+                        // Copy the signature file
+                        transfer(exportZip, entry, zos);
+                    }
+                });
+            }
+        }
```
Comment (bakajstep) [inline spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java:138]: Could we extract this into some methods? Just to be more readable for future.

**Review by bakajstep — COMMENTED** [review]
It would be good if someone with more experience also look at this PR a give it review.

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import org.candlepin.spec.bootstrap.data.builder.Products;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+@SpecTest
+public class ImportUpdateSinglePoolTest {
```
[the remark this answers is not in the document]
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+import java.io.InputStream;
+import java.time.OffsetDateTime;
+import java.util.Iterator;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipOutputStream;
+
+@OnlyInStandalone
+public class ImportWarningTest {
```
[the remark this answers is not in the document]
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        assertThat(subs)
+            .hasSize(2);
+
+        adminClient.consumers().unbindByEntitlementId(consumer.getUuid(), ent1.getId());
+        Thread.sleep(1000);
+
+        export = exportGenerator.export();
+        importJob = adminClient.owners().importManifestAsync(ownerKey, List.of(), export.file());
+        importJob = adminClient.jobs().waitForJob(importJob);
+        assertEquals("FINISHED", importJob.getState());
```
[the remark this answers is not in the document]
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java]: Oh nice! I didn't realize that we added this. Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        List<EntitlementDTO> ents = adminClient.consumers().listEntitlements(consumer.getUuid());
+        assertThat(ents).singleElement();
+        EntitlementDTO ent1 = ents.get(0);
+
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        String ownerKey = owner.getKey();
+
+        AsyncJobStatusDTO importJob = adminClient.owners().importManifestAsync(ownerKey, List.of(), export.file());
+        importJob = adminClient.jobs().waitForJob(importJob);
+        assertEquals("FINISHED", importJob.getState());
```
[the remark this answers is not in the document]
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            .minimal()
+            .withProduct(Products.random())
+            .withProduct(Products.random())
+            .export();
+
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+
+        AsyncJobStatusDTO importJob = adminClient.owners().importManifestAsync(owner.getKey(), List.of(), export.file());
+        importJob = adminClient.jobs().waitForJob(importJob);
+        assertEquals("FINISHED", importJob.getState());
```
[the remark this answers is not in the document]
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java (hunk 1 above)
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java (hunk 2 above)
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportUpdateSinglePoolTest.java]: Updated

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java (hunk 3 above)
Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningTest.java]: That is a good idea! I created `copyConsumerExportAndReplaceEnt` helper method and moved the logic that creates a copy of the consumer_export.zip file in the Export and replaced the entitlement in that copy.

File: spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningSpecTest.java [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+     * Converts the provided entitlement to Json. Also ensures that date fields in the {@link EntitlementDTO}
+     * and nested objects will not run into deserialization issues.
+     *
+     * @param ent
+     *  the entitlement to convert to json
+     *
+     * @return
+     *  the bytes of the converted Json string
+     */
+    private byte[] getEntitlementJson(EntitlementDTO ent) {
```
Comment (nikosmoum) [inline spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningSpecTest.java:147]: Why is this needed? Could we not just serialize it directly with the jackson API like `writeValueAsBytes`?

Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningSpecTest.java:147]: I tried using the `writeValueAsByte` method on the ObjectMapper, but it looks like we still run into the issue where the date fields are converted into what looks like the epoch value `"created":1668437865`. This causes issues being deserialized on in Candlepin, including the error message below. The `getEntitlementJson` gets around this issue by using the toString of the OffsetDateTime as the created/updated values: `"created":"2022-11-14T14:50:27Z"`

Candlepin Error message:
```
org.candlepin.async.JobExecutionException: Unable to extract export archive
        at org.candlepin.async.tasks.ImportJob.execute(ImportJob.java:245)
        at org.candlepin.async.JobManager.executeJob(JobManager.java:1273)
        at org.candlepin.async.JobMessageReceiver$MessageListener.handleMessage(JobMessageReceiver.java:350)
        at org.candlepin.messaging.impl.artemis.ArtemisConsumer$ArtemisMessageForwarder.onMessage(ArtemisConsumer.java:59)
        at org.apache.activemq.artemis.core.client.impl.ClientConsumerImpl.callOnMessage(ClientConsumerImpl.java:1001)
        at org.apache.activemq.artemis.core.client.impl.ClientConsumerImpl$Runner.run(ClientConsumerImpl.java:1154)
        at org.apache.activemq.artemis.utils.actors.OrderedExecutor.doTask(OrderedExecutor.java:56)
        at org.apache.activemq.artemis.utils.actors.OrderedExecutor.doTask(OrderedExecutor.java:31)
        at org.apache.activemq.artemis.utils.actors.ProcessorBase.executePendingTasks(ProcessorBase.java:67)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.apache.activemq.artemis.utils.ActiveMQThreadFactory$1.run(ActiveMQThreadFactory.java:118)
Caused by: org.candlepin.sync.ImportExtractionException: Unable to extract export archive
        at org.candlepin.sync.Importer.doExport(Importer.java:452)
        at org.candlepin.sync.Importer.loadStoredExport(Importer.java:209)
        at org.candlepin.controller.ManifestManager.importStoredManifest(ManifestManager.java:220)
        at com.google.inject.persist.jpa.JpaLocalTxnInterceptor.invoke(JpaLocalTxnInterceptor.java:64)
        at org.candlepin.async.tasks.ImportJob.execute(ImportJob.java:229)
        ... 11 common frames omitted
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `java.util.Date` from Floating-point value (token `JsonToken.VALUE_NUMBER_FLOAT`)
 at [Source: (FileReader); line: 1, column: 12] (through reference chain: org.candlepin.dto.manifest.v1.EntitlementDTO["created"])

```

Comment (JoshAlbrecht23) [reply spec-tests/src/test/java/org/candlepin/spec/imports/ImportWarningSpecTest.java]: Taking a closer look a this, we can disable the SerializationFeature.WRITE_DATES_AS_TIMESTAMPS on the ObjectMapper and that prevents the conversion of the OffsetDateTime to a timestamp value and allows use to use the `writeValueAsBytes` and get rid of the `getEntitlementJson` method!

**Review by nikosmoum — APPROVED** [review]
